pkg/apis/flowcontrol/validation/validation.go

```diff
@@ -0,0 +1,259 @@
+	}
+	if hasWildcard(rule.APIGroups) && len(rule.APIGroups) > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("apiGroups"), rule.APIGroups, "if '*' is present, must not specify other api groups"))
+	}
+	if hasWildcard(rule.Resources) && len(rule.Resources) > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("resources"), rule.Resources, "if '*' is present, must not specify other resources"))
+	}
+	if !supportedVerbs.IsSuperset(sets.NewString(rule.Verbs...)) {
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("verbs"), rule.Verbs, supportedVerbs.List()))
+	}
+
+	return allErrs
+}
+
+func ValidateFlowSchemaStatus(status *flowcontrol.FlowSchemaStatus, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	// conditions will not be validated
+	return allErrs
+}
+
+func ValidatePriorityLevelConfiguration(pl *flowcontrol.PriorityLevelConfiguration) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&pl.ObjectMeta, false, ValidatePriorityLevelConfigurationName, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidatePriorityLevelConfigurationSpec(&pl.Spec, pl.Name, field.NewPath("spec"))...)
+	allErrs = append(allErrs, ValidatePriorityLevelConfigurationStatus(&pl.Status, field.NewPath("status"))...)
+	return allErrs
+}
+
+func ValidatePriorityLevelConfigurationSpec(spec *flowcontrol.PriorityLevelConfigurationSpec, name string, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if name != flowcontrol.PriorityLevelConfigurationNameSystemTop && spec.Exempt {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("exempt"), "must not be exempt"))
+	}
+	if name != flowcontrol.PriorityLevelConfigurationNameWorkloadLow && spec.GlobalDefault {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("globalDefault"), "must not be global default"))
+	}
+	if spec.AssuredConcurrencyShares <= 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("assuredConcurrencyShares"), spec.AssuredConcurrencyShares, "must be positive"))
+	}
+	if spec.QueueLengthLimit <= 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("queueLengthLimit"), spec.QueueLengthLimit, "must be positive"))
+	}
+	if spec.HandSize < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("handSize"), spec.HandSize, "must be positive"))
+	}
+	if spec.Queues <= 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("queues"), spec.Queues, "must be positive"))
+	}
+	if spec.HandSize > spec.Queues {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("handSize"), spec.HandSize,
+			fmt.Sprintf("should not be greater than queues (%d)", spec.Queues)))
+	}
+	if !validateShuffleShardingParameters(spec.HandSize, spec.Queues) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("handSize"), spec.HandSize,
+			fmt.Sprintf("falling factorial of handSize (%d) and queues (%d) exceeds %d bits", spec.HandSize, spec.Queues, maxHashBits)))
```

Comment:
I thought we agreed to use an approximate test rather than exact falling factorial.

Reply:
i think we can address that in another pull #80710?